### PR TITLE
Improve the performance of tree listing

### PR DESCRIPTION
**Pull request by typeless**
Add two query strings `start` and `count` for the home page of repository.  If the values are omitted or invalid they are set to 0 and the total number of rows respectively.
   
For #502

### CODE REVIEW

File: routers/repo/view.go [hunk 1]
```diff
@@ -40,12 +41,33 @@ func renderDirectory(ctx *context.Context, treeLink string) {
 		return
 	}
 
+	query := ctx.Req.URL.Query()
+
+	var start, count int = 0, 100
```
Comment (lunny) [inline routers/repo/view.go]: I prefer the count could be configured and If we ignored some files we have to give a hint on the UI.

File: routers/repo/view.go [hunk 2]
```diff
@@ -40,12 +41,33 @@ func renderDirectory(ctx *context.Context, treeLink string) {
...
+
+	if _start, ok := query["start"]; ok && len(_start) == 1 {
```
Comment (lunny) [inline routers/repo/view.go]: `ctx. QueryInt ("start")` is more convenient.

File: routers/repo/view.go [hunk 3]
```diff
@@ -40,12 +41,33 @@ func renderDirectory(ctx *context.Context, treeLink string) {
...
+		start, err = strconv.Atoi(_start[0])
+		if err != nil {
+			ctx.Handle(500, "QueryString", err)
+			return
+		}
+	}
+
+	if _count, ok := query["count"]; ok && len(_count) == 1 {
```
Comment (lunny) [inline routers/repo/view.go]: `ctx.QueryInt ("count")`

File: routers/repo/view.go [hunk 4]
```diff
@@ -40,12 +41,33 @@ func renderDirectory(ctx *context.Context, treeLink string) {
...
+	}
+
 	entries, err := tree.ListEntries()
 	if err != nil {
 		ctx.Handle(500, "ListEntries", err)
 		return
 	}
-	entries.Sort()
+
+	entries = entries[start : start+count]
```
Comment (lunny) [inline routers/repo/view.go]: lost sort?

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/repo/view.go (hunk 4 above)
Comment (typeless) [reply routers/repo/view.go]: According to ftp://www.kernel.org/pub/software/scm/git/docs/git-mktree.html, that has been done by git itself when creating the tree object. And my trivial testing seems to confirm it.

File: routers/repo/view.go (hunk 1 above)
Comment (typeless) [reply routers/repo/view.go]: That would be better. Let me figure out how to do it properly.

**Comment by Bwko** [comment]
When count is not set it shows no commits. When coun/startt > commits Gitea panics

**Comment by typeless** [comment]
Please mark this as work-in-progress, I will finish this PR later (adding front-end controls as @lunny suggested).

@Bwko Thanks for pointing it out.

Edit: Or perhaps I can set the default `count` to the total number of rows, so it can still work as before and have follow-up PRs of UI supports in the future.

Edit2: The latest commit should fix the bug found by @Bwko

**Comment by Bwko** [comment]
@typeless Good work :+1:  Only the sorting is alphabetically and not folder first.

**Comment by typeless** [comment]
@Bwko My bad 😆 , didn't notice the nuances.

File: routers/repo/view.go [hunk 5]
```diff
@@ -12,6 +12,9 @@ import (
 	"path"
 	"strings"
```
Comment (appleboy) [inline routers/repo/view.go]: Remove redundant empty line.

**Review by appleboy — CHANGES_REQUESTED** [review]

File: routers/repo/view.go [hunk 6]
```diff
@@ -45,8 +45,19 @@ func renderDirectory(ctx *context.Context, treeLink string) {
...
 		return
 	}
+
 	entries.Sort()
 
+	start := ctx.QueryInt("start")
+	count := ctx.QueryInt("count")
+
+	// Check invalid values
+	if (start == 0 && count == 0) || start+count < start || start+count > len(entries) {
```
Comment (ethantkoenig) [inline routers/repo/view.go]: Could we replace `start+count < start` with `count < 0`?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@typeless Is this PR ready to review?

**Comment by typeless** [comment]
@lunny The UI part is still missing if that is desired to be included in this PR.

**Comment by bkcsoft** [comment]
@typeless UI would be required yes, otherwise there may be issue in the future :( Would be nice to have it fetch and render the next set of files via API, like GitHub has `load more entries...` when there's to many files 🙂

**Comment by bkcsoft** [comment]
You could also just include the `base/paginate.tmpl`, but then you need to use `?page=0` and a static count 🙁

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@typeless Is this completed?

**Comment by typeless** [comment]
@lunny I don't have a good idea about how the UI design should be done, before starting to get familiar with Semantic UI. Please bear with me as my spare timeslices for OSS projects are very limited recently.

If someone is willing to help implement the web UI (I am really not good at it), that would be appreciated.

**Comment by lunny** [comment]
OK. Please give the permissions to let maintainers to push to your branch, it's in this PR right sidebar. And I will move this to v1.2.

**Comment by typeless** [comment]
@lunny  Thanks. I have the "allow edits from maintainers" checked.

**Comment by bkcsoft** [comment]
@typeless @lunny I can look at implementing lazy-loading during the weekend 😄

**Comment by lunny** [comment]
@bkcsoft any update?

**Comment by bkcsoft** [comment]
@lunny I've been swamped with work unfortunately 🙁

**Pull request closed without merging**